File: data/src/main/java/org/cryptomator/data/db/dao/VaultDao.java
package org.cryptomator.data.db.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import org.cryptomator.data.db.entities.VaultEntity;
import org.cryptomator.data.db.entities.VaultWithCloud;

import java.util.List;

@Dao
public interface VaultDao {

	@Transaction
	@Query("SELECT * FROM VAULT_ENTITY")
	List<VaultWithCloud> loadAll();

	@Transaction
	@Query("SELECT * FROM VAULT_ENTITY WHERE _id = :id")
	VaultWithCloud load(Long id);

	@Insert
	long insert(VaultEntity entity);

	@Update
	void update(VaultEntity entity);

	@Delete
	void delete(VaultEntity entity);

	@Transaction
	default VaultWithCloud store(VaultWithCloud entity) {
		VaultEntity vault = entity.getVault();
		Long id = vault.getId();
		if (id == null) {
			id = insert(vault);
		} else {
			update(vault);
		}
		return load(id);
	}

	@Transaction
	default void delete(VaultWithCloud entity) {
		delete(entity.getVault());
	}
}

File: data/src/main/java/org/cryptomator/data/db/VaultDao.kt
package org.cryptomator.data.db

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import org.cryptomator.data.db.entities.VaultEntity

@Dao
interface VaultDao {

	@Query("SELECT * FROM VAULT_ENTITY WHERE id = :id LIMIT 1")
	fun load(id: Long): VaultEntity

	@Query("SELECT * from VAULT_ENTITY")
	fun loadAll(): List<VaultEntity>

	@Insert(onConflict = OnConflictStrategy.REPLACE)
	fun storeReplacing(entity: VaultEntity): RowId

	@Query("SELECT * FROM VAULT_ENTITY WHERE rowid = :rowId")
	fun loadFromRowId(rowId: RowId): VaultEntity

	@Transaction
	fun storeReplacingAndReload(entity: VaultEntity): VaultEntity {
		return loadFromRowId(storeReplacing(entity))
	}

	@Delete
	fun delete(entity: VaultEntity)
}

internal class DelegatingVaultDao(private val database: Invalidatable<CryptomatorDatabase>) : VaultDao {

	private val delegate: VaultDao
		get() = database.call().vaultDao()

	override fun load(id: Long): VaultEntity = delegate.load(id)

	override fun loadAll(): List<VaultEntity> = delegate.loadAll()

	override fun storeReplacing(entity: VaultEntity): RowId = delegate.storeReplacing(entity)

	override fun loadFromRowId(rowId: RowId): VaultEntity = delegate.loadFromRowId(rowId)

	override fun storeReplacingAndReload(entity: VaultEntity): VaultEntity = delegate.storeReplacingAndReload(entity)

	override fun delete(entity: VaultEntity) = delegate.delete(entity)
}